### PR TITLE
feat(keys-rotation): manage cookie secret key

### LINKED ISSUE #15362: [SIP-70] Secret keys rotations and stop using single SECRET_KEY
### In short
* Instead of using one master key aka flask.secret_key use different key for each use case
* support key rotation

### In long
Our organization security policies require to meet constraints in regard to using secret keys and credentials
* Our products should use a different key for each use so in case a specific key is stolen, all other objects are kept safe.
* Key can be changed or rotated at any time, so fresh keys can be maintained all the time.
* When Superset is crashed or exited, the keys cannot be lost so encrypted data could be decrypted.

## Proposed Change

* Each library/module which uses the master key by accessing app.secret_key will access the key by calling a get key method
* Use cases consisting of encryption-decryption flows will define methods handling procedures of rotations, means decrypt with old keys and re-encryption with a new key 
* Adding endpoints to declare rotations, they will trigger the declared procedures. 
* Easy way to define new dedicated keys in the future features 

## Guidelines
* Use feature flag
* Incremantaly by phases, each phase can consist of a series of PRs, PR for each specific secret-key use 
* Default values will be the current master key
* Can be customized for each use-case 
* Contribute to depended libraries 

## Plan
P1 - adapting required infrastructure without any usage yet
P2 - migrate using the new infrastructure
P3 - adding infrastructure to support key-changed declarations 
P4 - upgrade the dependent libraries that adapt the same mechanism 

### P1 - https://github.com/apache/superset/pull/15423
* Adapting the https://pypi.org/project/keys-management/ package 
   https://github.com/nielsen-oss/keys-management
* defining feature flag  

### P2 
for each dedicated key: 
 - add temporary code to inject secret key method instead of value - based on inheritance the current client's classes
 - define specific configurations for the specific key while the default value will be the master key
 - if necessary defining rotation methods 
 * review phase four why I said "temporary code"

#### P2 implementations
* Dedicated key for singing cookies: https://github.com/apache/superset/pull/15296
* Dedicated key for encryption-decryption insert and fetching "databases" rows (models)
* Dedicated key for Oauth login flow
* Dedicated key for Fab login and decorators

### P3
will be incrementally too, first adding REST API endpoints, then refactor or add more safe and robust way
such as CLI - 
first, support only an offline mode. 
Then build a small CLI socket-based service to support key rotations in runtime 

### P4
We plan to contribute the same mechanism directly to the dependent libraries.
for each adapted dependent library 
- remove the temporary code added in phase three 
- upgrade the library

### PR DESCRIPTION
**Pull request by ofekisr**
### SUMMARY
#15362 P2 implementaion, based on #15423

### CODE REVIEW

**Comment by ofekisr** [comment]
#15362

File: superset/config.py
Comment (amitmiran137) [inline superset/config.py:1229]:
```suggestion
# You could define it as lambda to load on every usage and change on the fly upon env var update like the following
# COOKIE_SIGNING_KEY = lambda: os.environ.get("COOKIE_SIGNING_KEY", SECRET_KEY)
COOKIE_SIGNING_KEY = os.environ.get("COOKIE_SIGNING_KEY", SECRET_KEY)
```

**Comment by ktmud** [comment]
Please add PR description

File: superset/initialization/secret_keys/__init__.py
Comment (craig-rueda) [inline superset/initialization/secret_keys/__init__.py:1]: Let's keep __init__ empty.

**Review by craig-rueda — COMMENTED** [review]
Is this part of SIP 70? Should we wait for approval before this is merged?

File: superset/app.py
Comment (craig-rueda) [inline superset/app.py:109]: Not sure I agree with this pattern of hanging fields off of the App. A more widely accepted/established pattern is to use a custom `extension` that can be added to the Flask app via the `extensions`. We already have a mechanism for this (extensions.py)

File: superset/initialization/secret_keys/__init__.py
Comment (craig-rueda) [inline superset/initialization/secret_keys/__init__.py:27]: Do we need this?

**Review by craig-rueda — CHANGES_REQUESTED** [review]
Also wondering if there's some overlap with the encrypted field manager I added a while back. I'd prefer if we merged the functionality in order to prevent duplication.

Lastly, I don't see any tests.

**Comment by ofekisr** [comment]
> Also wondering if there's some overlap with the encrypted field manager I added a while back. I'd prefer if we merged the functionality in order to prevent duplication.
> 
> Lastly, I don't see any tests.

what you added is a specific concern - encryptedColumnFactory 
here is totally different concern

**Comment by craig-rueda** [comment]
> > Also wondering if there's some overlap with the encrypted field manager I added a while back. I'd prefer if we merged the functionality in order to prevent duplication.
> > Lastly, I don't see any tests.
> 
> what you added is a specific concern - encryptedColumnFactory
> here is totally different concern

Sorta - They both deal with keys and the source of key material. I think it would be great to have a central secrets manager that can be easily overridden by orgs that want to extend functionality. Also, we have established patterns for this kinda stuff (app factory pattern) that should be followed here.

**Pull request closed without merging**
